File: sample/src/main/java/com/xload/endpointsample/java/HomeActivity.java
package com.xload.endpointsample.java;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.xload.endpointsample.Constants;
import com.xload.endpointsample.R;
import com.xload.endpointsforandroid.api.models.LinkedWallet;
import com.xload.endpointsforandroid.api.models.WalletStatus;
import com.xload.endpointsforandroid.modules.XLD;
import com.xload.endpointsforandroid.utils.OnXLDListener;

import static java.sql.DriverManager.println;

/**
 * @author John Paul Cas
 * Copyright (c) 2020 XLD Tech Labs. All rights reserved.
 */
public class HomeActivity extends AppCompatActivity {

    private String xldUserId = null;

    private ProgressBar startLoading;
    private Button btnGetConversion;
    private Button btnGetWalletStatus;
    private Button btnLinkAccount;
    private LinearLayout linkWalletContainer;
    private LinearLayout llAppStatus;
    private TextView tvMessage;

    private EditText etAppId;
    private EditText etPartnerId;
    private EditText etWalletAddress;
    private EditText etOtp;
    private EditText etStatusUserId;
    private EditText etConversionAppId;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);
        startLoading = findViewById(R.id.startLoading);
        btnGetConversion = findViewById(R.id.btnGetConversion);
        btnGetWalletStatus = findViewById(R.id.btnGetWalletStatus);
        btnLinkAccount = findViewById(R.id.btnLinkAccount);
        linkWalletContainer = findViewById(R.id.linkWalletContainer);
        llAppStatus = findViewById(R.id.llAppStatus);
        tvMessage = findViewById(R.id.tvMessage);

        etAppId = findViewById(R.id.etAppId);
        etStatusUserId = findViewById(R.id.etStatusUserId);
        etPartnerId = findViewById(R.id.etPartnerId);
        etWalletAddress = findViewById(R.id.etWalletAddress);
        etOtp = findViewById(R.id.etOtp);
        etConversionAppId = findViewById(R.id.etConversionAppId);

        setClickListener();
    }

    private void setClickListener() {
        // set sample data
        etAppId.setText(Constants.APP_ID);
        etPartnerId.setText(Constants.SAMPLE_PARTNER_ID);
        etWalletAddress.setText(Constants.SAMPLE_WALLET_ADDRESS);
        etOtp.setText(String.valueOf(Constants.SAMPLE_OTP));

        // Get conversion sample
        btnGetConversion.setOnClickListener(v -> {
            getConversion();
        });

        /**
         * Get wallet status
         */
        btnGetWalletStatus.setOnClickListener(v -> {
           getWalletStatus();
        });

        btnLinkAccount.setOnClickListener(v -> {
            linkedAccount();
        });
    }

    /**
     * Linked wallet account
     */
    private void linkedAccount() {
        String appId = etAppId.getText().toString();
        String walletAddress = etWalletAddress.getText().toString();
        int otp = Integer.parseInt(etOtp.getText().toString());
        String partnerId = etPartnerId.getText().toString();

        if (etAppId.getText().toString().isEmpty()) {
            etWalletAddress.setError("App ID cannot be empty");
        } else if (etWalletAddress.getText().toString().isEmpty()) {
            etWalletAddress.setError("Wallet address cannot be empty");
        } else if (etOtp.getText().toString().isEmpty()) {
            etOtp.setError("OTP cannot be empty");
        } else if (etPartnerId.getText().toString().isEmpty()) {
            etPartnerId.setError("Partner user id cannot be empty");
        } else {
            submitLinkedWallet();
            XLD.getInstance(
                    Constants.KEY,
                    Constants.SECRET,
                    true /*development mode*/
            ).linkWallet(
                    appId,
                    walletAddress,
                    otp,
                    partnerId,
                    new OnXLDListener<LinkedWallet>() {
                        @Override
                        public void loading(boolean isLoading) {
                        }

                        @Override
                        public void error(@org.jetbrains.annotations.Nullable String error) {
                            startLoading.setVisibility(View.GONE);
                            showMessage(error);
                        }

                        @Override
                        public void success(@org.jetbrains.annotations.Nullable LinkedWallet result) {
                            xldUserId = result.getXldUserId();

                            etStatusUserId.setText(result.getXldUserId());
                            etConversionAppId.setText(result.getAppId());

                            startLoading.setVisibility(View.GONE);
                            llAppStatus.setVisibility(View.VISIBLE);

                            String response = new StringBuilder("Wallet\n")
                                    .append("appId=").append(result.getAppId()).append("\n")
                                    .append("isLinked=").append(result.getLinked()).append("\n")
                                    .append("otp=").append(result.getOtp()).append("\n")
                                    .append("partnerId=").append(result.getPartnerUserId()).append("\n")
                                    .append("walletAddress=").append(result.getWalletAddress()).append("\n")
                                    .append("xldUserId=").append(result.getXldUserId()).append("\n")
                                    .toString();
                            showMessage(response);
                        }
                    }
            );
        }
    }

    /**
     * Get wallet status
     */
    private void getWalletStatus() {
        if (etStatusUserId.getText().toString().isEmpty()) {
            etStatusUserId.setError("XLD User ID cannot be empty.");
            return;
        } else {
            startLoading.setVisibility(View.VISIBLE);

            String userId = etStatusUserId.getText().toString();
            XLD.getInstance(
                    Constants.KEY,
                    Constants.SECRET,
                    true /*development mode*/
            ).getWalletStatus(
                    userId,
                    new OnXLDListener<WalletStatus>() {
                        @Override
                        public void loading(boolean isLoading) {

                        }

                        @Override
                        public void error(@org.jetbrains.annotations.Nullable String error) {
                            startLoading.setVisibility(View.GONE);
                            showMessage(error);
                        }

                        @Override
                        public void success(@org.jetbrains.annotations.Nullable WalletStatus result) {
                            startLoading.setVisibility(View.GONE);
                            String msg = new StringBuilder("Wallet status:\n")
                                    .append("isLinked=").append(result.getLinked()).append("\n")
                                    .append("partnerUserId=").append(result.getPartnerUserId()).append("\n")
                                    .append("xldUserId=").append(result.getXldUserId()).toString();
                            showMessage(msg);
                        }
                    }
            );
        }
    }

    private void getConversion() {
        if (etConversionAppId.getText().toString().isEmpty()) {
            etConversionAppId.setError("App ID cannot be empty.");
            return;
        } else {
            showGetConversionLoading();

            String appId = etConversionAppId.getText().toString();
            XLD.getInstance(
                    Constants.KEY,
                    Constants.SECRET,
                    true
            ).getConversion(appId, new OnXLDListener<Double>() {
                @Override
                public void loading(boolean isLoading) {

                }

                @Override
                public void error(@org.jetbrains.annotations.Nullable String error) {
                    startLoading.setVisibility(View.GONE);
                    showMessage(error);
                }

                @Override
                public void success(@org.jetbrains.annotations.Nullable Double result) {
                    startLoading.setVisibility(View.GONE);
                    String msg = new StringBuilder("Conversion\n")
                            .append("conversion=").append(result.toString())
                            .toString();
                    showMessage(msg);
                }
            });
        }
    }

    private void submitLinkedWallet() {
        llAppStatus.setVisibility(View.GONE);
        startLoading.setVisibility(View.VISIBLE);
    }

    private void showMessage(String msg) {
        tvMessage.setText(msg);
    }

    private void showGetConversionLoading() {
        startLoading.setVisibility(View.VISIBLE);
        if (etStatusUserId.getText().toString().isEmpty()) {
            llAppStatus.setVisibility(View.GONE);
        }
    }

}

File: sample/src/main/java/com/xload/endpointsample/HomeActivity.kt
package com.xload.endpointsample

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.xload.endpointsforandroid.api.models.LinkedWallet
import com.xload.endpointsforandroid.api.models.WalletStatus
import com.xload.endpointsforandroid.modules.XLD
import com.xload.endpointsforandroid.utils.OnXLDListener
import kotlinx.android.synthetic.main.activity_home.*

class HomeActivity : AppCompatActivity() {

    var xldUserId: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
        setClickListener()
    }

    private fun setClickListener() {
        // set sample data
        etConversionAppId.setText(Constants.APP_ID)
        etAppId.setText(Constants.APP_ID)
        etPartnerId.setText(Constants.SAMPLE_PARTNER_ID)
        etWalletAddress.setText(Constants.SAMPLE_WALLET_ADDRESS)
        etOtp.setText(Constants.SAMPLE_OTP.toString())

        /**
         * Get converstion sample
         */
        btnGetConversion.setOnClickListener {
            getConversion()
        }

        /**
         * Get wallet status
         */
        btnGetWalletStatus.setOnClickListener {
            getWalletStatus()
        }

        /**
         * Link account sample
         */
        btnLinkAccount.setOnClickListener {
           linkedAccount()
        }

    }

    /**
     * Linked wallet account
     */
    private fun linkedAccount() {
        val walletAddress = etWalletAddress.text.toString()
        val otp = etOtp.text.toString().toInt()
        val partnerId = etPartnerId.text.toString()
        val appId: String = etAppId.text.toString()

        when {
            etAppId.text.toString().isEmpty() -> {
                etAppId.setError("App ID cannot be empty")
                return
            }
            etWalletAddress.text.toString().isEmpty() -> {
                etWalletAddress.setError("Wallet address cannot be empty")
                return
            }
            etOtp.text.toString().isEmpty() -> {
                etOtp.setError("OTP cannot be empty")
                return
            }
            etPartnerId.text.isEmpty() -> {
                etPartnerId.setError("Partner user id cannot be empty")
                return
            }
            else -> {
                submitLinkedWallet()
                XLD.getInstance(Constants.KEY, Constants.SECRET, true).linkWallet(
                    appId = appId,
                    walletAddress = walletAddress,
                    otp = otp,
                    partnerUserId = partnerId,
                    listener = object : OnXLDListener<LinkedWallet> {
                        override fun loading(isLoading: Boolean) {
                        }

                        override fun error(error: String?) {
                            startLoading.visibility = View.GONE
                            showMessage(error ?: "")
                        }

                        override fun success(result: LinkedWallet?) {
                            println("DEBUG: LinkWallet = $result")
                            xldUserId = result?.xldUserId

                            etStatusUserId.setText(result?.xldUserId)
                            etConversionAppId.setText(result?.appId)

                            startLoading.visibility = View.GONE
                            llAppStatus.visibility = View.VISIBLE

                            val response = StringBuilder("Wallet\n")
                                .append("appId=${result?.appId}\n")
                                .append("isLinked=${result?.linked}\n")
                                .append("otp=${result?.otp}\n")
                                .append("partnerId=${result?.partnerUserId}\n")
                                .append("walletAddress=${result?.walletAddress}\n")
                                .append("xldUserId=${result?.xldUserId}\n")
                                .toString()
                            showMessage(response)
                        }

                    }
                )
            }
        }
    }

    /**
     * Get wallet status by app ID
     */
    private fun getWalletStatus() {
        if (etStatusUserId.text.toString().isEmpty()) {
            etStatusUserId.setError("XLD User ID cannot be empty.")
            return
        } else {
            startLoading.visibility = View.VISIBLE

            val userId = etStatusUserId.text.toString()
            XLD.getInstance(
                Constants.KEY,
                Constants.SECRET,
                true
            ).getWalletStatus(userId,
                object : OnXLDListener<WalletStatus> {
                    override fun loading(isLoading: Boolean) {
                    }

                    override fun error(error: String?) {
                        startLoading.visibility = View.GONE
                        showMessage(error ?: "")
                    }

                    override fun success(result: WalletStatus?) {
                        startLoading.visibility = View.GONE
                        val msg = StringBuilder("Wallet status:\n")
                            .append("isLinked=${result?.linked}\n")
                            .append("partnerUserId=${result?.partnerUserId}\n")
                            .append("xldUserId=${result?.xldUserId}\n")
                        showMessage(msg.toString())
                    }

                }
            )
        }
    }

    private fun getConversion() {
        if (etConversionAppId.text.toString().isEmpty()) {
            etConversionAppId.setError("App ID cannot be empty.")
            return
        } else {
            showGetConversionLoading()

            val appId = etConversionAppId.text.toString()
            XLD.getInstance(
                Constants.KEY,
                Constants.SECRET,
                true
            ).getConversion(appId, object : OnXLDListener<Double> {
                override fun loading(isLoading: Boolean) {
                }

                override fun error(error: String?) {
                    startLoading.visibility = View.GONE
                    showMessage(error ?: "")
                }

                override fun success(result: Double?) {
                    startLoading.visibility = View.VISIBLE
                    val result = StringBuilder("Conversion\n")
                        .append("conversion = ${result}")
                        .toString()
                    showMessage(result)
                }

            })
        }
    }

    private fun submitLinkedWallet() {
        llAppStatus.visibility = View.GONE
        startLoading.visibility = View.VISIBLE
    }

    private fun showMessage(msg: String) {
        tvMessage.setText(msg)
    }

    private fun showGetConversionLoading() {
        startLoading.visibility = View.VISIBLE
        if (etStatusUserId.text.toString().isEmpty()) {
            llAppStatus.visibility = View.GONE
        }
    }
}
